### PR TITLE
attendence functionality

### PR DESCRIPTION
**Pull request by RiyaSantore04**

### CODE REVIEW

File: app/controllers/admin_main/attendences_controller.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+module AdminMain
+  class AttendencesController < ApplicationController
+    before_action :set_attendance, only: %i[show edit]
+    def index
+      @attendences = Attendence.all
+      @attendences = Attendence.order('created_at desc').page(params[:page])
```
Comment (rorpyll) [inline app/controllers/admin_main/attendences_controller.rb:8]: why are you using  same code 2 times?

File: app/views/admin_main/attendences/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+          <th scope="col">Employee Name</th>
+          <th scope="col">Checkin Time</th>
+          <th scope="col">Checkout Time</th>
+          <th scope="col">Working Hours</th>
+        </tr>
+      </thead>
+
+      <tbody class="search-data">
+        <% array = [] %>
+        <% @attendences.all.each do |attendence| %>
```
Comment (rorpyll) [inline app/views/admin_main/attendences/index.html.erb:19]: why are you using attendances.all?

File: app/views/shared/_header.html.erb [hunk 3]
```diff
@@ -84,45 +84,80 @@
           <% end %>
           <% end %>
       </li>
-      
+
+
       <li class="nav-item">
-          <%= link_to employees_path , class: "nav-link collapsed" do %>
+          <% if current_employee.role.name == "Admin" %>
```
Comment (rorpyll) [inline app/views/shared/_header.html.erb:90]: use instance method for that

File: app/views/shared/_header.html.erb [hunk 4]
```diff
@@ -84,45 +84,80 @@
...
-              <%= 'Employees' %>
           <% end %>
           <% end %>
       </li>
+      
 
+      <li class="nav-item">
+          <%= link_to employees_path , class: "nav-link collapsed" do %>
+          <i class="bi bi-person"></i>
+            <% unless current_employee.role.name == "HR" || current_employee.role.name == "Admin"%>
```
Comment (rorpyll) [inline app/views/shared/_header.html.erb:102]: use instance method for that

File: app/views/shared/_header.html.erb [hunk 5]
```diff
@@ -84,45 +84,80 @@
...
+      
       
       <li class="nav-item">
           <%= link_to  performances_path, class: "nav-link collapsed" do %>
           <i class="bi bi-bar-chart"></i>
           <%= 'Performance' %>
           <% end %>
       </li>
+
+    <% if current_employee.role.name == "Admin"%>
```
Comment (rorpyll) [inline app/views/shared/_header.html.erb:118]: same

File: app/views/shared/_header.html.erb [hunk 6]
```diff
@@ -84,45 +84,80 @@
...
+    <% end %>
+
       <li class="nav-item">
           <%= link_to  leafs_path, class: "nav-link collapsed" do %>
           <i class="bi bi-menu-button-wide"></i>
           <%= 'Leave' %>
           <% end %>
       </li>
+   
+       <% if current_employee.role.name == "Admin"%>
```
Comment (rorpyll) [inline app/views/shared/_header.html.erb:142]: same change

File: app/views/shared/_header.html.erb [hunk 7]
```diff
@@ -135,12 +170,24 @@
...
           <% end %>
       </li>
-      <li class="nav-item">
-          <%= link_to  tickets_path, class: "nav-link collapsed" do %>
-          <i class="bi bi-menu-button-wide"></i>
-          <%= 'Ticket' %>
-          <% end %>
-      </li>
+ 
+      <% if current_employee.role.name == "Admin"%>
```
Comment (rorpyll) [inline app/views/shared/_header.html.erb:174]: same

File: config/routes.rb [hunk 8]
```diff
@@ -35,10 +35,25 @@
...
+    
     resources :attendences
     get '/emp-attendance/:id', to: 'attendences#show_attendence', as: 'show_attendance'
     get '/search', to: 'attendences#search'
   end
+
+  namespace :admin_main do
+    resources :salaries
+    resources :attendences
+    resources :tickets
```
Comment (rorpyll) [inline config/routes.rb:48]: why code is repeating

File: spec/controllers/admin_main/attendence_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+  let!(:designation) { FactoryBot.create(:designation) }
+  let!(:employee) { FactoryBot.create(:employee, role_id: roles(:role_one).id, designation_id: designation.id) }
+  let!(:attendence) { FactoryBot.create(:attendence, employee_id: employee.id, checkin_time: Time.now) }
+
+  before do
+    sign_in(employee)
+  end
+
+  describe 'GET #index' do
+    it 'should return all customers' do
```
Comment (rorpyll) [inline spec/controllers/admin_main/attendence_controller_spec.rb:15]: write proper sentence

File: spec/controllers/admin_main/attendence_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+      get :index
+      expect(response).to have_http_status(:success)
```
Comment (rorpyll) [inline spec/controllers/admin_main/attendence_controller_spec.rb:17]: expect with object

File: spec/controllers/admin_main/attendence_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    it 'return root path when attendence not exit' do
+      get :show, params: { id: '' }
+      expect(response.status).to eq(302)
+    end
+  end
+
+  describe '#edit' do
+    it 'response successfully when attendence exit' do
+      get :edit, params: { id: attendence.id }
+      expect(assigns(:attendence)).to eq(attendence)
```
Comment (rorpyll) [inline spec/controllers/admin_main/attendence_controller_spec.rb:37]: your edit action not return any object, how you  can compare with object?

File: spec/controllers/admin_main/attendence_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+    it 'return root path when attendence not exit' do
+      get :edit, params: { id: '' }
+      expect(response.status).to eq(302)
+    end
+  end
+
+  describe '#update' do
+    it 'response successfully when attendence exit' do
+      patch :update, params: { id: attendence.id }
```
Comment (rorpyll) [inline spec/controllers/admin_main/attendence_controller_spec.rb:48]: what are you updating here?

File: spec/requests/admin_main/attendences_spec.rb [hunk 13]
```diff
@@ -0,0 +1,7 @@
+require 'rails_helper'
```
Comment (rorpyll) [inline spec/requests/admin_main/attendences_spec.rb:1]: why are you add this blank file

File: spec/controllers/admin_main/attendence_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+      expect(assigns(:attendence)).to eq(attendence)
+    end
+
+    it 'return root path when attendence not exit' do
+      patch :update, params: { id: '' }
+      expect(response.status).to eq(302)
+    end
+  end
```
Comment (rorpyll) [inline spec/controllers/admin_main/attendence_controller_spec.rb:56]: write spec for not update condition

**Review by rorpyll — CHANGES_REQUESTED** [review]

File: app/views/admin_main/attendences/index.html.erb (hunk 2 above)
Comment (RiyaSantore04) [reply app/views/admin_main/attendences/index.html.erb:19]: Changes done

File: app/views/shared/_header.html.erb (hunk 3 above)
Comment (RiyaSantore04) [reply app/views/shared/_header.html.erb:90]: Changes done

File: app/views/shared/_header.html.erb (hunk 4 above)
Comment (RiyaSantore04) [reply app/views/shared/_header.html.erb:102]: Changes done

File: app/views/shared/_header.html.erb (hunk 5 above)
Comment (RiyaSantore04) [reply app/views/shared/_header.html.erb:118]: Changes done

File: app/views/shared/_header.html.erb (hunk 6 above)
Comment (RiyaSantore04) [reply app/views/shared/_header.html.erb:142]: Changes done

File: app/views/shared/_header.html.erb (hunk 7 above)
Comment (RiyaSantore04) [reply app/views/shared/_header.html.erb:174]: Changes done

File: spec/controllers/admin_main/attendence_controller_spec.rb (hunk 9 above)
Comment (RiyaSantore04) [reply spec/controllers/admin_main/attendence_controller_spec.rb:15]: Changes done

File: spec/requests/admin_main/attendences_spec.rb (hunk 13 above)
Comment (RiyaSantore04) [reply spec/requests/admin_main/attendences_spec.rb:1]: Done by mistake

File: spec/controllers/admin_main/attendence_controller_spec.rb (hunk 14 above)
Comment (RiyaSantore04) [reply spec/controllers/admin_main/attendence_controller_spec.rb:56]: Changes done

File: spec/controllers/admin_main/attendence_controller_spec.rb (hunk 12 above)
Comment (RiyaSantore04) [reply spec/controllers/admin_main/attendence_controller_spec.rb:48]: Changes done

File: spec/controllers/admin_main/attendence_controller_spec.rb (hunk 10 above)
Comment (RiyaSantore04) [reply spec/controllers/admin_main/attendence_controller_spec.rb:17]: Changes done

**Pull request closed without merging**